File: test/conformance/host-fixtures/presence.java
// Host fixture implementation for the host-export-presence conformance
// test (host-module-abi D6). The JVM lane synthesizes the declared host
// class (Config) as a shared $DealRt record carrying the canonical
// externals identity (@$external/host.presence/Config); the
// Config_defaults map is the preserved defaults-map seam — it carries
// ONLY port (8080), exactly like the Lua fixture, so a provided
// fallback/peers name raises E8007 at construction
// (host-class-extra-field pins the rejection).
//
// ISSUE-0303 (jvm-v12-host-abi-completion D4): the harness references
// the synthesized record class, never a nested JDK mirror bound to the
// host implementation.
//
// The top-level class is deliberately package-private: the corpus file
// keeps its <name>.java stem while the JVM lane deploys the source
// beside the default-package artifacts under the classNameFor name
// Host<Name>.java (test/JvmConformanceTest.java HOST_JAVA pattern).
final class HostPresence {
  public static final java.util.Map<String, Object> Config_defaults =
      java.util.Map.of("port", Integer.valueOf(8080));

  public static Object ping() {
    return "pong";
  }
}

File: test/conformance/host-fixtures/presence.js
"use strict";

// Host fixture implementation for the host-export-presence conformance
// test (host-module-abi D6). The declared class export carries the
// canonical externals identity descriptor (@$external/host/presence/Config) and a
// <C>_defaults table — runtime construction through the synthesized
// class symbol depends on both. The preserved defaults-map seam
// (host-module-abi D2, ISSUE-0331): the loader passes the defaults
// table through verbatim, and this host does NOT mark its absent
// optional fields with $rt.MISSING — so a provided declared optional
// absent from Config_defaults raises E8007 at construction (the
// host-class-extra-field fixture).
module.exports = {
  ping: function () {
    return "pong";
  },

  Config: {
    $kind: "class",
    $classname: "@$external/host/presence/Config",
  },

  Config_defaults: {
    port: 8080,
  },
};
